base/timing.jl

```diff
@@ -115,35 +124,39 @@ function format_bytes(bytes) # also used by InteractiveUtils
     end
 end
 
-function time_print(elapsedtime, bytes=0, gctime=0, allocs=0, compile_time=0)
-    timestr = Ryu.writefixed(Float64(elapsedtime/1e9), 6)
-    length(timestr) < 10 && print(" "^(10 - length(timestr)))
-    print(timestr, " seconds")
-    parens = bytes != 0 || allocs != 0 || gctime > 0 || compile_time > 0
-    parens && print(" (")
-    if bytes != 0 || allocs != 0
-        allocs, ma = prettyprint_getunits(allocs, length(_cnt_units), Int64(1000))
-        if ma == 1
-            print(Int(allocs), _cnt_units[ma], allocs==1 ? " allocation: " : " allocations: ")
-        else
-            print(Ryu.writefixed(Float64(allocs), 2), _cnt_units[ma], " allocations: ")
-        end
-        print(format_bytes(bytes))
-    end
-    if gctime > 0
+const print_lock = ReentrantLock()
+function time_print(elapsedtime, bytes=0, gctime=0, allocs=0, compile_time=0, newline=false)
+    lock(print_lock) do
+        timestr = Ryu.writefixed(Float64(elapsedtime/1e9), 6)
+        length(timestr) < 10 && print(" "^(10 - length(timestr)))
+        print(timestr, " seconds")
+        parens = bytes != 0 || allocs != 0 || gctime > 0 || compile_time > 0
+        parens && print(" (")
         if bytes != 0 || allocs != 0
-            print(", ")
+            allocs, ma = prettyprint_getunits(allocs, length(_cnt_units), Int64(1000))
+            if ma == 1
+                print(Int(allocs), _cnt_units[ma], allocs==1 ? " allocation: " : " allocations: ")
+            else
+                print(Ryu.writefixed(Float64(allocs), 2), _cnt_units[ma], " allocations: ")
+            end
+            print(format_bytes(bytes))
         end
-        print(Ryu.writefixed(Float64(100*gctime/elapsedtime), 2), "% gc time")
-    end
-    if compile_time > 0
-        if bytes != 0 || allocs != 0 || gctime > 0
-            print(", ")
+        if gctime > 0
+            if bytes != 0 || allocs != 0
+                print(", ")
+            end
+            print(Ryu.writefixed(Float64(100*gctime/elapsedtime), 2), "% gc time")
+        end
+        if compile_time > 0
+            if bytes != 0 || allocs != 0 || gctime > 0
+                print(", ")
+            end
+            print(Ryu.writefixed(Float64(100*compile_time/elapsedtime), 2), "% compilation time")
         end
-        print(Ryu.writefixed(Float64(100*compile_time/elapsedtime), 2), "% compilation time")
+        parens && print(")")
+        newline && println()
+        nothing
     end
-    parens && print(")")
-    nothing
```

Comment:
The diff is deceptively messy here. I just wrapped `time_print`'s code in a print lock, and added the `newline` arg to print a newline within the print lock at the end

Reply:
`timev_print` still prints its own newline outside the lock. If it's important for the newline to be printed inside the lock, we should do it always. If it's not, we should do it never and the caller can print a newline if wanted.

Reply:
For `timev_print` I've switched to using the internal newline print. From what I can tell `time_print` needs optionally do both it's used during the julia build process in `sysimage.jl`